File: java/853_Car_Fleet.java
class Car{
    int leftDis;
    double time;
    Car(int p, double t){
        this.leftDis = p;
        this.time = t;
    }
}

class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int nCars = position.length;
        Car[] cars = new Car[nCars];
        for(int i = 0 ; i < nCars ; i++){
            cars[i] = new Car( (target-position[i]) , (double)(target-position[i])/speed[i] );
        }
        Arrays.sort(cars, new Comparator<Car>(){
            @Override
            public int compare(Car c1, Car c2){
                return c1.leftDis-c2.leftDis;
            }
        } );


        int res = 0, t = 0;
        double slowest = 0.0;

        while (t < nCars) {
            if (cars[t].time > slowest){
                res++; //if cars[t] arrives sooner, it can't be caught
                slowest = cars[t].time;
            }
            t++;
        }

        return res; //lone car is fleet (if it exists)
    }
}

File: python/853_Car_Fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = [(target-pos,spd) for (pos,spd) in zip(position,speed)]
        cars = sorted(cars)
        print(cars)
        
        fleet = 0
        slowest = 0
        for i in range(len(cars)):
            time = cars[i][0] / cars[i][1]
            if time > slowest:
                fleet += 1
                slowest = time
        
        return fleet